posts/loggingPython.md

```diff
@@ -0,0 +1,325 @@
+```
+
+When capturing an exception, it’s essential to add context. Similar to when you’re debugging, you want to know who the user is and what they were doing when the exception struck.
+
+```python
+import sys
+
+# remember to set up your logger
+
+def captureException():
+    logger.warning(sys.exc_info())
+```
+
+![](./images/loggingPython/exception1.png)
+
+This gives us a `traceback` object that we can use to see the stack from our exception. We didn’t even need to pass anything into our `captureException` method.
+
+Let’s figure out what the `traceback` object gives us and clean up this information a little.
+
+```python
+import sys
+import traceback
+
+def captureException():
+    r = list(sys.exc_info())
+
+    e = dict()
+    e["name"] = r[1]
+```
+
+![](./images/loggingPython/exception2.png)
+
+Now, this can be cleaned up using some string manipulation in Python.
+
+## Storing & Accessing These Logs
+
+Now that you've learned to write these (beautiful) logs, you've got to learn what to do with them. I mean it would be a waste to write them to the console and leave them to be deleted when the program closes.
+
+### Logging to a File
+
+Generally, you don't (just) want to log to the console for them to be deleted immediately. The logging module makes it easy to write your logs to a local file using a `handler`.
+
+```python
+import logging
+logger = logging.getLogger(__name__)
+
+handler = logging.FileHandler('myLogs.log')
+handler.setLevel(logging.INFO)
+
+logger.addHandler(handler)
+logger.info('You can find this written in myLogs.log')
+```
+
+### Logging to the Cloud
+
+Writing your logs to the cloud seriously makes your life easier, it serves as a layer of abstraction so you don’t have to worry how they are getting to the service and you can focus your time on what’s important.
+
+![Timber](./images/loggingPython/timber.png)
+
+_Disclaimer: I’m a current employee @ Timber. This section is entirely optional, but I sincerely believe that logging to the cloud will make your life easier (and you can try it for completely free)._
+
+```bash
+pip install timber
+```
+
+```python
+import logging
+import timber
+
+logger = logging.getLogger(__name__)
+
+timber_handler = timber.TimberHandler(api_key='...')
+logger.addHandler(timber_handler)
+```
+
+**That’s it.** All you need to do is get your API key from [timber.io](https://timber.io/), and you’ll be able to see your logs. We automatically capture them from the logging module, so you can continue to follow best-practices and log normally, while we seamlessly add context.
+
+Now that you’ve created _gorgeous_ logs that give the next developer a view of the user’s actions, you need to know best practices for searching and reacting to these logs.
+
+### Searching
+
+Going through your logs from a file is a tedious process, it’s honestly better to use a [cloud-based solution](https://timber.io/) that acts as a layer of abstraction, so you don’t have to think about parsing huge log files.
```

Comment:
I'd remove the comment that it's tedious. Also, `grep critical log_file_name` is a lot better than writing a Python script.